src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,35 +5,104 @@
-    : base_t(std::move(layout)), IBackendObject(std::move(core::smart_refctd_ptr<const ILogicalDevice>(pool->getOriginDevice()))), m_version(0ull), m_pool(std::move(pool)), m_poolOffset(poolOffset), m_descriptorStorageOffsets(std::move(offsets))
+IGPUDescriptorSet::IGPUDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout, core::smart_refctd_ptr<IDescriptorPool>&& pool, IDescriptorPool::SStorageOffsets&& offsets)
+    : base_t(std::move(layout)), IBackendObject(std::move(core::smart_refctd_ptr<const ILogicalDevice>(pool->getOriginDevice()))), m_version(0ull), m_pool(std::move(pool)), m_storageOffsets(std::move(offsets))
 {
     for (auto i = 0u; i < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
     {
         // There is no descriptor of such type in the set.
-        if (m_descriptorStorageOffsets.data[i] == ~0u)
+        if (m_storageOffsets.data[i] == ~0u)
             continue;
 
         const auto type = static_cast<asset::IDescriptor::E_TYPE>(i);
 
         // Default-construct the core::smart_refctd_ptr<IDescriptor>s because even if the user didn't update the descriptor set with ILogicalDevice::updateDescriptorSet we
         // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
-        std::uninitialized_default_construct_n(m_pool->getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
+        std::uninitialized_default_construct_n(m_pool->getDescriptorStorage(type) + m_storageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
     }
 
     const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
     if (mutableSamplerCount > 0)
-        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
+        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_storageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
 }
 
 IGPUDescriptorSet::~IGPUDescriptorSet()
 {
     if (!isZombie())
     {
-        m_pool->deleteSetStorage(this);
-        if (m_pool->m_creationParameters.flags.hasFlags(IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT))
-            m_pool->m_descriptorSetAllocator.free_addr(m_poolOffset, 1);
+        auto dummy = this;
+        m_pool->deleteSetStorage(dummy);
     }
 }
 
+bool IGPUDescriptorSet::processWrite(const IGPUDescriptorSet::SWriteDescriptorSet& write)
+{
+    assert(write.dstSet == this);
+
+    system::ILogger* logger = nullptr;
+    {
+        auto debugCallback = getOriginDevice()->getPhysicalDevice()->getDebugCallback();
+        if (debugCallback)
+            logger = debugCallback->getLogger();
+    }
```

Comment:
how about we grab the logger in the constructor of the pool (or accept it from the outside as a constructor parameter) ?

Reply:
Can't we just add a `getLogger` to `ILogicalDevice` which returns a `logger_opt_ptr`?

Reply:
we could, but its still faster to keep a `logger_opt_ptr` here as a member so we dont need to retrieve it on every DS write or copy

Reply:
Right, I'll just grab it from `ILogicalDevice` in pool's ctor.

Reply:
Resolved.